lls/cache.c

```diff
@@ -0,0 +1,428 @@
+{
+	int32_t ret = rte_hash_add_key(cache->hash, ip_key);
+	if (unlikely(ret == -EINVAL)) {
+		RTE_LOG(ERR, HASH, "Invalid params, %s entry not added\n",
+			cache->name);
+		rte_mempool_put(cache->entry_pool, new_entry);
+	} else if (unlikely(ret == -ENOSPC)) {
+		RTE_LOG(ERR, HASH, "No space in cache, %s entry not added\n",
+			cache->name);
+		rte_mempool_put(cache->entry_pool, new_entry);
+	} else {
+		struct lls_cache_entry *existing_entry = cache->data[ret];
+		if (existing_entry != NULL) {
+			/*
+			 * Update existing entry. This can only be a valid
+			 * entry updating an older entry, but cannot be an
+			 * invalid entry updating an older entry. Invalid
+			 * entries just serve as reference counting
+			 * placeholders, so if there's already an entry in
+			 * the table, __lls_cache_refcnt() would have
+			 * recognized it and just updated its refcnt.
+			 *
+			 * Take the characteristics from the new entry.
+			 * There is a race condition on the hardware address,
+			 * as it could be being updated while a functional
+			 * block uses it. However, most of the time the update
+			 * will just be a refresh and the address will not
+			 * actually change, and even if it is changing it is
+			 * expected that it could take time for an LL address
+			 * change to correctly propagate.
+			 */
+			ether_addr_copy(&new_entry->ha, &existing_entry->ha);
+			existing_entry->ts = new_entry->ts;
+			existing_entry->port_id = new_entry->port_id;
+			existing_entry->stale = new_entry->stale;
+			existing_entry->valid = new_entry->valid;
+			rte_mempool_put(cache->entry_pool, new_entry);
+		} else {
+			/* Add new entry (valid or invalid) to the cache. */
+			cache->data[ret] = new_entry;
+		}
+
+		if (get_lls_conf()->debug)
+			lls_cache_dump(cache);
+	}
+}
+
+static void
+__lls_cache_del(struct lls_cache *cache, const void *ip_key)
+{
+	int32_t ret = rte_hash_del_key(cache->hash, ip_key);
+	if (unlikely(ret == -ENOENT))
+		RTE_LOG(ERR, HASH, "No entry found, %s entry not deleted\n",
+			cache->name);
+	else if (unlikely(ret == -EINVAL))
+		RTE_LOG(ERR, HASH, "Invalid params, %s entry not deleted\n",
+			cache->name);
+	else {
+		if (unlikely(cache->data[ret]->refcnt > 0))
+			RTE_LOG(ERR, GATEKEEPER,
+				"Positive interest, %s entry not deleted\n",
+				cache->name);
+		else {
+			rte_mempool_put(cache->entry_pool, cache->data[ret]);
+			cache->data[ret] = NULL;
+			if (get_lls_conf()->debug)
+				lls_cache_dump(cache);
+		}
+	}
+}
+
+static void
+__lls_cache_refcnt(struct lls_config *lls_conf, struct lls_cache *cache,
+	const void *ip_key, struct lls_cache_entry *entry, enum lls_cache_op op)
+{
+	int ret = rte_hash_lookup(cache->hash, ip_key);
+	if (ret == -ENOENT) {
+		if (op == LLS_CACHE_INC) {
+			/* Add placeholder entry and try to resolve it. */
+			__lls_cache_add(cache, ip_key, entry);
+			cache->xmit_req(lls_conf->net->front.id, ip_key, NULL);
+			cache->xmit_req(lls_conf->net->back.id, ip_key, NULL);
```

Comment:
Deciding whether or not we should send a request on a particular interface gets tricky among these three things:

- Making sure ARP is enabled on the interface.
- Making sure the interface is enabled itself (in the case of the back interface).
- Making sure we're only sending out on an interface that matches the network mask.

I've centralized all of this into one function now (`send_request()` in lls/cache.c), and added a `TODO` tag for the network mask part.